### PR TITLE
Pre authenticated user

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
- supported pre-authenticated user

### How it does work:
Api Gateway fetches users from Backend to resolve save's user for OAuth2 and Basic.
Api Gateway saves id, name and role from resolved save's user in headers.
Backend (and Sandbox) using authentication-service module resolves `PreAuthenticatedAuthenticationToken` from values in headers without checking database.

It's part of #2336

### CODE REVIEW

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/configs/ApiSwaggerSupport.kt [hunk 1]
```diff
@@ -27,12 +29,24 @@ import io.swagger.v3.oas.annotations.security.SecurityScheme
 annotation class ApiSwaggerSupport
 
 /**
- * Indicates that an operation requires `X-Authorization-Source` header to be set
+ * Indicates that an operation requires `X-Authorization-*` headers to be set
  */
 @Parameter(
```
Comment (nulls) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/configs/ApiSwaggerSupport.kt:34]: these headers look are useless for me.
I think it should be removed since user cannot provide it (as `X-Authorization-Source` after migration to k8s), they are populated by api-gateway.

This change leads to big changes in source code and open api. Will do it in separate PR

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/utils/SecurityTestingUtils.kt [hunk 2]
```diff
@@ -4,14 +4,19 @@
...
 import org.springframework.security.core.context.SecurityContextHolder
 
 /**
  * @param action mutation for a mocked user in test security context
  */
-internal fun mutateMockedUser(action: UsernamePasswordAuthenticationToken.() -> Unit) {
+internal fun mutateMockedUser(id: Long, name: String = "N/A") {
     SecurityContextHolder.getContext().apply {
-        authentication = (authentication as UsernamePasswordAuthenticationToken).apply(action)
+        authentication = AuthenticationUserDetails(
```
Comment (nulls) [inline save-backend/src/test/kotlin/com/saveourtool/save/backend/utils/SecurityTestingUtils.kt:16]: currently we do:
```kotlin
@WithMockUser(name = "something", roles = ["SINGLE_ROLE"])
fun test() {
    mutateMockedUser(id = 99)
    TODO()
}
```

It can be refactored to:
```kotlin
fun test() {
    mutateMockedUser(id = 99, name = "something", roles = "SINGLE_ROLE")
    TODO()
}
```

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/AuthenticationUserDetails.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    companion object {
+        @Suppress("GENERIC_VARIABLE_WRONG_DECLARATION")
+        private val log = getLogger<AuthenticationUserDetails>()
+        private const val NO_CREDENTIALS = "N/A"
+
+        /**
+         * @return [AuthenticationUserDetails] created from values in headers
+         */
+        fun ServerWebExchange.toAuthenticationUserDetails(): AuthenticationUserDetails? {
```
Comment (nulls) [inline authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/AuthenticationUserDetails.kt:59]: can be covered by test

Comment (nulls) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/AuthenticationUserDetails.kt:71]: done

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import reactor.kotlin.core.publisher.toMono
+import java.security.Principal
+
+/**
+ * Filter, that mutate existing exchange,
+ * inserts user's info into Authorization headers instead of existing value, not paying attention to the credentials,
+ * since at this moment they are already checked by gateway.
+ */
+@Component
+class AuthorizationHeadersGatewayFilterFactory(
```
Comment (nulls) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt:26]: it's to hard to cover by test, will be done in separate task in scope of #2336

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt [hunk 5]
```diff
@@ -22,42 +21,35 @@ import reactor.core.publisher.Mono
...
         .retrieve()
         .onStatus({ it.is4xxClientError }) {
             Mono.error(ResponseStatusException(it.statusCode()))
         }
-        .toEntity<String>()
-        .map {
-            springUserDetailsReader.readValue(it.body)
+        .toEntity<AuthenticationUserDetails>()
+        .flatMap { responseEntity ->
+            responseEntity.body.toMono().orNotFound { "Body is empty" }
```
Comment (kgevorkyan) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:52]: maybe 
```suggestion
            responseEntity.body.toMono().orNotFound { "Authentication body is empty" }
```

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                    }
+                    .build()
+            }
+            .defaultIfEmpty(exchange)
+            .flatMap { chain.filter(it) }
+    }
+
+    private fun resolveSaveUser(principal: Principal): Mono<AuthenticationUserDetails> = when (principal) {
+        is OAuth2AuthenticationToken -> backendService.findByOriginalLogin(principal.authorizedClientRegistrationId, principal.name)
+        is UsernamePasswordAuthenticationToken -> (principal.principal as? AuthenticationUserDetails)
```
[the remark this answers is not in the document]
Comment (nulls) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt:48]: we expect here only `uthenticationUserDetails` and even we throw exception here. I think it can be safely changed to `as`.

about what we discussed, it was the statement like `when (value) { is A -> value as A }`. when type is matched in section `is` and then in body

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/config/WebSecurityConfig.kt [hunk 7]
```diff
@@ -53,9 +67,9 @@ class WebSecurityConfig(
             // FixMe: Properly support CSRF protection https://github.com/saveourtool/save-cloud/issues/34
             csrf().disable()
         }
-        .httpBasic {
-            it.authenticationManager(authenticationManager)
-        }
+        .httpBasic()
+        .disable()
```
[the remark this answers is not in the document]
Comment (nulls) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/config/WebSecurityConfig.kt:71]: we don't need basic auth (or even any spring's auth) in backend and sandbox -- we use pre-authenticated user in internal communication.
if request goes through api-gateway, it's authenticated.

[the remark this answers is not in the document]
Comment (nulls) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/config/WebSecurityConfig.kt:71]: actually, httpBasic is not enabled by default. Wil remove `.httpBasic().disable()`

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    private val backendService: BackendService,
+) : AbstractGatewayFilterFactory<Any>() {
+    override fun apply(config: Any?): GatewayFilter = GatewayFilter { exchange: ServerWebExchange, chain: GatewayFilterChain ->
+        exchange.getPrincipal<Principal>()
+            .flatMap { resolveSaveUser(it) }
+            .map { user ->
+                exchange.mutate()
+                    .request { builder ->
+                        builder.headers { headers: HttpHeaders ->
+                            headers.remove(HttpHeaders.AUTHORIZATION)
```
Comment (akuleshov7) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt:36]: btw - why do you remove authorization headers? 
May be not create it initially for particular endpoints?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt [hunk 9]
```diff
@@ -68,8 +61,8 @@ class UsersController(
     @GetMapping("/find-by-name/{userName}")
     fun findByName(
         @PathVariable userName: String,
-    ): Mono<StringResponse> = userService.findByName(userName).map {
-        ResponseEntity.ok().body(springUserDetailsWriter.writeValueAsString(it))
+    ): Mono<AuthenticationUserDetailsResponse> = userService.findByName(userName).map {
+        ResponseEntity.ok().body(AuthenticationUserDetails(it))
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt:65]: Ah, you moved it from a simple string to some data class, that's why you need new UserDetails for authorization? (this is how I get it)

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/configs/ApiSwaggerSupport.kt (hunk 1 above)
Comment (akuleshov7) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/configs/ApiSwaggerSupport.kt:34]: yes, definitely, we are isolated with k8s

**Review by akuleshov7 — APPROVED** [review]

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt (hunk 8 above)
Comment (nulls) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/utils/AuthorizationHeadersGatewayFilterFactory.kt:36]: we don't use authorization header in backend. I just cleanup it to be sure that we don't pass through this value from external request

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/AuthenticationUserDetails.kt [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+import org.springframework.security.core.authority.AuthorityUtils
+import org.springframework.security.core.userdetails.UserDetails
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+
+/**
+ * @property id [com.saveourtool.save.entities.User.id]
+ * @property name [com.saveourtool.save.entities.User.name]
+ * @property role [com.saveourtool.save.entities.User.role]
+ */
+data class AuthenticationUserDetails(
```
[the remark this answers is not in the document]
Comment (nulls) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/AuthenticationUserDetails.kt:18]: we have `AuthenticationDetails` which contains `id` and it's set to `Authentication.details`.
But `Authentication.principal` can be any object and Spring supports `UserDetails` to resolve `Authentication.name`.
I decided to store all required info pre-authenticated user in principal and `UserDetails` interface allows to pass username to Spring engine

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt (hunk 9 above)
Comment (nulls) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/internal/UsersController.kt:65]: yes, but here is another change:
I don't why but we didn't use the Spring engine to parse Object provided by backend to api-gateway. We created a custom `objectMapper` to decode it from String.
I guess the reason is in `getters`. Even in current implementation I have to add `@JsonIgnore` in `AuthenticationUserDetails`. Previously we used `org.springframework.security.core.userdetails.User` and additionally registered `org.springframework.security.jackson2.SecurityJackson2Modules` (before it we have a custom extension for `SpringUser` and a custom serializer for it)
